### PR TITLE
new 3rd party driver for Astromechanics Canon Lens Controller

### PR DESCRIPTION
**Pull request by chkettu**
An indi 3rd party driver for Astromechanics Lens Control (Focusser for Canon lenses).

https://www.astromechanics.org/ascom.html
Find the documentation attached to this merge request. Feedback, improvements, etc. welcome.

[Astromechanics_FOC_Documentation.tar.gz](https://github.com/indilib/indi/files/3130854/Astromechanics_FOC_Documentation.tar.gz)

### CODE REVIEW

**Comment by chkettu** [comment]
I've no idea what's wrong with that build system. It builds with docker (for example: docker run --rm -it jochym/indi-bionic ./run-build.sh chkettu pull)...

Any advice needed!

File: 3rdparty/CMakeLists.txt [hunk 1]
```diff
@@ -393,6 +394,11 @@ if (WITH_MGEN)
 add_subdirectory(indi-mgen)
 endif (WITH_MGEN)
 
+# Astromechanics Focuser
+if (WITH_ASTROMECHFOC)
+add_subdirectory(indi_astromechfoc)
```
Comment (knro) [inline 3rdparty/CMakeLists.txt]: indi_astromechfoc --> indi-astromechfoc

File: 3rdparty/indi-astromechfoc/astromech_focuser.cpp [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+{
+    Astromechanics_foc->ISSnoopDevice(root);
+}
+
+/************************************************************************************
+ *
+************************************************************************************/
+astromechanics_foc::astromechanics_foc()
+{
+    setVersion(0, 1);
```
Comment (knro) [inline 3rdparty/indi-astromechfoc/astromech_focuser.cpp]: version should be from the CMakeLists.txt file, see other drivers (INDI_ASTROMECHFOC_VERSION_MAJOR)

File: 3rdparty/indi-astromechfoc/astromech_focuser.cpp [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    FI::SetCapability(FOCUSER_CAN_ABS_MOVE | FOCUSER_CAN_REL_MOVE);
+}
+
+/************************************************************************************
+ *
+************************************************************************************/
+const char *astromechanics_foc::getDefaultName()
+{
+    return (const char *)"Astromechanics FOC";
```
Comment (knro) [inline 3rdparty/indi-astromechfoc/astromech_focuser.cpp]: (const char *) can be dropped.

File: 3rdparty/indi-astromechfoc/astromech_focuser.cpp [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+}
+
+/************************************************************************************
+ *
+************************************************************************************/
+void astromechanics_foc::ISGetProperties(const char *dev)
```
Comment (knro) [inline 3rdparty/indi-astromechfoc/astromech_focuser.cpp]: ISGetProperties is not used so better remove the whole function

File: 3rdparty/indi-astromechfoc/astromech_focuser.cpp [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+************************************************************************************/
+bool astromechanics_foc::Handshake()
+{
+    char FOC_cmd[32] = "P#";
+    char FOC_res[32] = {0};
+    int FOC_pos_measd = 0;
+    int nbytes_written = 0;
+    int nbytes_read = 0;
+
+    LOGF_INFO("Handshake", 0);
```
Comment (knro) [inline 3rdparty/indi-astromechfoc/astromech_focuser.cpp]: This should be LOG_DEBUG

File: 3rdparty/indi-astromechfoc/astromech_focuser.cpp [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    if (CanAbsMove()) {
+        LOGF_INFO("CanAbsMove",0);
```
Comment (knro) [inline 3rdparty/indi-astromechfoc/astromech_focuser.cpp]: No need to pass 0, you can just say LOG_INFO("CanAbsMove")

But this is for the user, so better have a better message like LOG_INFO("Focuser supports absolute motion");

File: 3rdparty/indi-astromechfoc/astromech_focuser.cpp [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    } else {
+        LOGF_INFO("Can't ABSMOVE",0);
+    }
+    tty_write_string(PortFD, FOC_cmd, &nbytes_written);
+    LOGF_INFO("CMD (%s)", FOC_cmd);
```
Comment (knro) [inline 3rdparty/indi-astromechfoc/astromech_focuser.cpp]: INFO --> DEBUG

We have INFO, WARN, ERROR, and DEBUG

File: 3rdparty/indi-astromechfoc/astromech_focuser.cpp [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+
+    SetApperture(0);
+
+    return false;
+}
+
+/************************************************************************************
+ *
+************************************************************************************/
+bool astromechanics_foc::ISNewSwitch(const char *dev, const char *name, ISState *states, char *names[], int n)
```
Comment (knro) [inline 3rdparty/indi-astromechfoc/astromech_focuser.cpp]: ISNewSwitch is not used, so better drop it.

File: 3rdparty/indi-astromechfoc/astromech_focuser.cpp [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+    char FOC_cmd[32]  = "M";
+    char abs_pos_char[32]  = {0};
+    int nbytes_written = 0;
+
+    //int pos = GetAbsFocuserPosition();
+    sprintf(abs_pos_char, "%d", targetTicks);
+    strcat(abs_pos_char, "#");
+    strcat(FOC_cmd, abs_pos_char);
+
+    LOGF_INFO("CMD (%s)", FOC_cmd);
```
Comment (knro) [inline 3rdparty/indi-astromechfoc/astromech_focuser.cpp]: INFO --> DEBUG

File: 3rdparty/indi-astromechfoc/astromech_focuser.cpp [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+    // Calculation of the demand absolute position
+    uint32_t targetTicks = FocusAbsPosN[0].value + (ticks * (dir == FOCUS_INWARD ? -1 : 1));
+    FocusAbsPosNP.s = IPS_BUSY;
+    IDSetNumber(&FocusAbsPosNP, nullptr);
+
+    return MoveAbsFocuser(targetTicks);
+}
+
+
+bool astromechanics_foc::saveConfigItems(FILE *fp)
```
Comment (knro) [inline 3rdparty/indi-astromechfoc/astromech_focuser.cpp]: Not used, drop it.

File: 3rdparty/indi-astromechfoc/astromech_focuser.cpp (hunk 6 above)
Comment (chkettu) [reply 3rdparty/indi-astromechfoc/astromech_focuser.cpp]: Removed that - was for development purposes only.

**Comment by knro** [comment]
Documentation added: https://indilib.org/devices/focusers/astromechanics-canon-lens-controller.html

I also added Nightly and Stable releases of the driver to the PPA.

**Comment by knro** [comment]
Can you please check this post? https://www.indilib.org/forum/general/5182-i-can-t-find-the-astromechanics-focuser-driver.html#39129

**Comment by chkettu** [comment]
checked and replied to the post.

**Comment by knro** [comment]
@chkettu Can you please check this post? https://indilib.org/forum/focusers-filter-wheels/5534-astromechanics-focuser.html

Any ideas?

**Pull request merged**